src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs

```diff
@@ -197,6 +172,90 @@ public sealed override void Initialize(AnalysisContext analysisContext)
                         context.ReportDiagnostic(type.CreateDiagnostic(Rule, type.FormatMemberName()));
                     }
                 });
+
+                return;
+
+                // Local functions
+
+                bool CanBeCoClassAttributeContext(INamedTypeSymbol type)
+                     => coClassAttributeSymbol != null && type.TypeKind == TypeKind.Interface;
+
+                INamedTypeSymbol? FindTypeIfCoClassAttribute(AttributeData attribute)
+                {
+                    RoslynDebug.Assert(coClassAttributeSymbol != null);
+
+                    if (attribute.AttributeClass.Equals(coClassAttributeSymbol) &&
+                        attribute.ConstructorArguments.Length == 1 &&
+                        attribute.ConstructorArguments[0].Kind == TypedConstantKind.Type &&
+                        attribute.ConstructorArguments[0].Value is INamedTypeSymbol typeSymbol &&
+                        typeSymbol.TypeKind == TypeKind.Class)
+                    {
+                        return typeSymbol;
+                    }
+
+                    return null;
+                }
+
+                bool CanBeDesignerAttributeContext(INamedTypeSymbol type)
+                    => designerAttributeSymbol != null && (type.TypeKind == TypeKind.Interface || type.TypeKind == TypeKind.Class);
+
+                INamedTypeSymbol? FindTypeIfDesignerAttribute(AttributeData attribute, Compilation compilation)
+                {
+                    RoslynDebug.Assert(designerAttributeSymbol != null);
+
+                    if ((attribute.ConstructorArguments.Length != 1 && attribute.ConstructorArguments.Length != 2) ||
+                        !attribute.AttributeClass.Equals(designerAttributeSymbol))
+                    {
+                        return null;
+                    }
+
+                    switch (attribute.ConstructorArguments[0].Value)
```

Comment:
Can this cause out of range exception?

Reply:
We have previously check there is 1 or 2 args so IMO it should not.

Reply:
Yes, but it's `||`ed with `attribute.AttributeClass.Equals(designerAttributeSymbol)`.

If, for some reason, the runtime in a future version introduced a parameterless constructor, would this cause a problem?
Probably the other check needs to be `&&` instead? Or I might be missing something.

Reply:
If there is a number of args different from 1 and 2 -> bail-out. If it is not the right attribute -> bail-out. So for me there is no problem when reaching this point we are guarantee to have 1 or 2 params (so 0 exists).

If a parameterless constructor is introduced, then it won't match the 1st condition (i.e. 1 or 1 params) so that's safe too.

Reply:
@Evangelink Oh yes my mistake.